core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -124,162 +136,170 @@ static Builder buildFrom(ViewMetadata base) {
 
   class Builder {
     private static final int LAST_ADDED = -1;
+    private final List<ViewVersion> versions;
+    private final List<Schema> schemas;
+    private final List<ViewHistoryEntry> history;
+    private final Map<String, String> properties;
+    private final List<MetadataUpdate> changes;
     private int formatVersion = DEFAULT_VIEW_FORMAT_VERSION;
-    private String location;
-    private List<Schema> schemas = Lists.newArrayList();
     private int currentVersionId;
-    private Integer lastAddedVersionId;
-    private Integer lastAddedSchemaId;
-    private List<ViewVersion> versions = Lists.newArrayList();
-    private List<ViewHistoryEntry> history = Lists.newArrayList();
-    private Map<String, String> properties = Maps.newHashMap();
-    private List<MetadataUpdate> changes = Lists.newArrayList();
+    private String location;
 
-    private Builder() {}
+    // internal change tracking
+    private Integer lastAddedVersionId = null;
+
+    // indexes
+    private final Map<Integer, ViewVersion> versionsById;
+    private final Map<Integer, Schema> schemasById;
+
+    private Builder() {
+      this.versions = Lists.newArrayList();
+      this.versionsById = Maps.newHashMap();
+      this.schemas = Lists.newArrayList();
+      this.schemasById = Maps.newHashMap();
+      this.history = Lists.newArrayList();
+      this.properties = Maps.newHashMap();
+      this.changes = Lists.newArrayList();
+    }
 
     private Builder(ViewMetadata base) {
-      this.formatVersion = base.formatVersion();
-      this.location = base.location();
-      this.schemas = Lists.newArrayList(base.schemas());
-      this.currentVersionId = base.currentVersionId();
       this.versions = Lists.newArrayList(base.versions());
+      this.versionsById = Maps.newHashMap(base.versionsById());
+      this.schemas = Lists.newArrayList(base.schemas());
+      this.schemasById = Maps.newHashMap(base.schemasById());
```

Comment:
Rather than running the index method each time a schema or version is added, this pulls the index from the last version and keeps it up to date.